### PR TITLE
Automatic Shift Transfer

### PR DESCRIPTION
**Pull request by ZephyrTFA**
## About The Pull Request

Adds in an automated way to have shifts end, if people vote for it to happen

## Why It's Good For The Game

Rounds probably shouldnt be as long as they are, and this also allows the server to automatically restart itself.

## Changelog
:cl:
add: Automatic Shift Transfer
/:cl:

### CODE REVIEW

**Review by atakiya — CHANGES_REQUESTED** [review]
Also move it out of `hrzn/` to `code/modules/` (and subsystem to the subsystem folder)
That folder is unsupported and will be removed in the future.

**Review by atakiya — APPROVED** [review]

**Comment by atakiya** [comment]
Also, CI failure is because of
`REASON #1: Automatic Transfer(/datum/controller/subsystem/auto_transfer) is a subsystem meant to initialize but doesn't get set as initialized.`

**Comment by ZephyrTFA** [comment]
> Also, CI failure is because of
> `REASON #1: Automatic Transfer(/datum/controller/subsystem/auto_transfer) is a subsystem meant to initialize but doesn't get set as initialized.`

Resolved

File: code/controllers/configuration/entries/general.dm [hunk 1]
```diff
@@ -521,3 +521,21 @@
...
+	default = "Automated Service System"
+
+/datum/config_entry/string/auto_transfer_vote_message
+	default = "End shift early?"
+
+/datum/config_entry/number/auto_transfer_first
+	default = 20
+
+/datum/config_entry/number/auto_transfer_concurrent
+	default = 10
```
Comment (MissFox0810) [inline code/controllers/configuration/entries/general.dm:541]: Can you add some comments on what each of these config entries do. (they may be self explanatory now, but for future)

**Review by MissFox0810 — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/auto_transfer.dm [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	for(var/c in GLOB.clients)
+		var/client/C = c
+		var/datum/action/vote/V = new
+		V.name = "Vote: [SSvote.question]"
+		C.player_details.player_actions += V
+		V.Grant(C.mob)
+		SSvote.generated_actions += V
+		if(C.prefs.toggles & SOUND_ANNOUNCEMENTS)
+			SEND_SOUND(C, sound('sound/misc/bloop.ogg'))
+	addtimer(CALLBACK(src, .proc/end_vote), vp)
```
Comment (Azarak) [inline code/controllers/subsystem/auto_transfer.dm:53]:
```suggestion
	var/vote_period = CONFIG_GET(number/vote_period)
	SSvote.time_remaining = round(vp/10)
	to_chat(world, "\n<span class='infoplain'><font color='purple'><b>[CONFIG_GET(string/auto_transfer_vote_message)]</b>\nType <b>vote</b> or click <a href='byond://winset?command=vote'>here</a> to place your votes.\nYou have [DisplayTimeText(vp)] to vote.</font></span>")
	for(var/client/voter_client as anything in GLOB.clients)
		var/datum/action/vote/vote = new
		vote .name = "Vote: [SSvote.question]"
		voter_client.player_details.player_actions += vote
		vote.Grant(voter_client.mob)
		SSvote.generated_actions += vote
		if(voter_client.prefs.toggles & SOUND_ANNOUNCEMENTS)
			SEND_SOUND(voter_client, sound('sound/misc/bloop.ogg'))
	addtimer(CALLBACK(src, .proc/end_vote), vote_period)
```

Comment (Azarak) [reply code/controllers/subsystem/auto_transfer.dm:53]: Descriptive variables and type casting in the for statement

Comment (ZephyrTFA) [reply code/controllers/subsystem/auto_transfer.dm:53]: Fair point; I just mostly copy/pasted this code from SSvote

File: code/controllers/subsystem/auto_transfer.dm [hunk 3]
```diff
@@ -0,0 +1,62 @@
+SUBSYSTEM_DEF(auto_transfer)
+	name = "Automatic Transfer"
+	flags = SS_NO_FIRE
+
+/datum/controller/subsystem/auto_transfer/Initialize(start_timeofday)
+	. = ..() // HEY IDIOT MAYBE CALL THE PARENT PROC YOU ARE REQUIRED TO CALL. YES YOU
```
Comment (atakiya) [inline code/controllers/subsystem/auto_transfer.dm:6]:
```suggestion
	. = ..()
```
unneeded comment?

**Review by atakiya — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/auto_transfer.dm [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+		return
+
+	SSvote.reset()
+	SSvote.choices = list("Yes", "No")
+	SSvote.question = CONFIG_GET(string/auto_transfer_vote_message)
+	SSvote.mode = "custom"
+	SSvote.initiator = CONFIG_GET(string/auto_transfer_announcer)
+	SSvote.started_time = world.time
+	var/vote_period = CONFIG_GET(number/vote_period)
+	SSvote.time_remaining = round(vp/10)
```
Comment (ZephyrTFA) [inline code/controllers/subsystem/auto_transfer.dm:42]:
```suggestion
	SSvote.time_remaining = round(vote_period/10)
```

File: code/controllers/subsystem/auto_transfer.dm [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	to_chat(world, "\n<span class='infoplain'><font color='purple'><b>[CONFIG_GET(string/auto_transfer_vote_message)]</b>\nType <b>vote</b> or click <a href='byond://winset?command=vote'>here</a> to place your votes.\nYou have [DisplayTimeText(vp)] to vote.</font></span>")
```
Comment (ZephyrTFA) [inline code/controllers/subsystem/auto_transfer.dm:43]:
```suggestion
	to_chat(world, "\n<span class='infoplain'><font color='purple'><b>[CONFIG_GET(string/auto_transfer_vote_message)]</b>\nType <b>vote</b> or click <a href='byond://winset?command=vote'>here</a> to place your votes.\nYou have [DisplayTimeText(vote_period)] to vote.</font></span>")
```

**Review by atakiya — APPROVED** [review]

**Pull request closed without merging**